### PR TITLE
API: add DataFrame.nunique() and DataFrameGroupBy.nunique()

### LINKED ISSUE #14336: API: add DataFrame.nunique() and DataFrameGroupBy.nunique()
When exploring a data set, I often need to `df.apply(pd.Series.nunique)` or `df.apply(lambda x: x.nunique())`. How about adding this as `nunique()`-method parallel to `DataFrame.count()` (`count` and `unique` are also the two most basic infos displayed by `DataFrame.describe()`)?

I think there are also use cases for this as a `groupby`-method, for example when checking a candidate primary key for different lines (values):

``` python
>>> import pandas as pd
>>> df = pd.DataFrame({'id': ['spam', 'eggs', 'eggs', 'spam'], 'value': [1, 5, 5, 2]})
>>> df.groupby('id').filter(lambda g: (g.apply(pd.Series.nunique) > 1).any())
     id  value
0  spam      1
3  spam      2
```

### PR DESCRIPTION
**Pull request by xflr6**
- [x] closes #14336
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
these are not going to be very efficient. Pls add some benchmarks.

**Comment by xflr6** [comment]
I could not get the benchmarks to run with `asv` under win32, so now there is something basically following what is already there for `count()`.

There is no 'fast-path' in the implementation of `nunique()` for `describe(include='all')` either, right?

**Comment by jreback** [comment]
can you rebase and post some asv numbers

**Comment by xflr6** [comment]
OK, rebased. None of the [proposed methods](http://pandas.pydata.org/pandas-docs/stable/contributing.html#running-the-performance-test-suite) to run `asv` on this seem to work on my win machine, e.g.:

```
(C:\Users\User\Miniconda2) C:\Users\User\Desktop\pandas\asv_bench>asv run 4c13e.
.9ab44 -b _nunique
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py2.7-Cython-libpython-matplotlib-numexpr-numpy-openp
yxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
·· Installing into conda-py2.7-Cython-libpython-matplotlib-numexpr-numpy-openpyx
l-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
· Running 3 total benchmarks (1 commits * 1 environments * 3 benchmarks)
[  0.00%] · For pandas commit hash 9ab44eec:
[  0.00%] ·· Building for conda-py2.7-Cython-libpython-matplotlib-numexpr-numpy-
openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  0.00%] ··· Error running C:\Users\User\Desktop\pandas\asv_bench\env\eebfa40b3
935b6d282712e13c756274b\python.exe setup.py build
              STDOUT -------->
[...]
              build\temp.win-amd64-2.7\Release\pandas\src\datetime\np_datetime_s
trings.o:np_datetime_strings.c:(.text+0x12): undefined reference to `localtime_r
'
              collect2.exe: error: ld returned 1 exit status
              error: command 'C:\\Users\\User\\Miniconda2\\Scripts\\gcc.bat' fai
led with exit status 1

[  0.00%] ·· Skipping conda-py2.7-Cython-libpython-matplotlib-numexpr-numpy-open
pyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 33.33%] ··· Benchmark groupby.groupby_nunique.time_groupby_nunique skipped
[ 66.67%] ··· Benchmark groupby.GroupBySuite.time_nunique skipped
[100.00%] ··· Benchmark frame_methods.frame_nunique.time_frame_nunique skipped
```

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -3899,6 +3899,21 @@ def count(self):
...
+        """
+        Return Series with number of distinct observations per group.
+
+        .. versionadded:: 0.20.0
+
+        Parameters
+        ----------
+        dropna : boolean, default True
+            Don't include NaN in the counts.
+        """
```
Comment (jreback) [inline pandas/core/groupby.py]: can you add a Returns and Examples sections

File: pandas/core/frame.py [hunk 2]
```diff
@@ -4947,6 +4947,27 @@ def f(x):
...
+        .. versionadded:: 0.20.0
+
+        Parameters
+        ----------
+        axis : {0 or 'index', 1 or 'columns'}, default 0
+            0 or 'index' for row-wise, 1 or 'columns' for column-wise
+        dropna : boolean, default True
+            Don't include NaN in the counts.
+
+        Returns
```
Comment (jreback) [inline pandas/core/frame.py]: can you add Examples section

File: pandas/tests/groupby/test_groupby.py [hunk 3]
```diff
@@ -2791,6 +2791,16 @@ def test_count_cross_type(self):  # GH8169
             result = df.groupby(['c', 'd']).count()
             tm.assert_frame_equal(result, expected)
 
+    def test_nunique(self):
+        df = DataFrame({
+            'A': list('abbacc'),
+            'B': list('abxacc'),
+            'C': list('abbacx'),
+        })
+        expected = DataFrame({'A': [1] * 3, 'B': [1, 2, 1], 'C': [1, 1, 2]})
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: test with both as_index=True and False

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -2791,6 +2791,16 @@ def test_count_cross_type(self):  # GH8169
...
+        result = df.groupby('A', as_index=False).nunique()
+        tm.assert_frame_equal(result, expected)
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: also can you test with dropna=True and False

**Comment by jreback** [comment]
@xflr6 ok lgtm. just some added tests / docs. ping on green.

don't worry about the benchmarks.

**Comment by xflr6** [comment]
Thanks, extended the docs and tests, CI passed.

File: pandas/core/frame.py [hunk 5]
```diff
@@ -4969,6 +4969,39 @@ def f(x):
...
+        -------
+        nunique : Series
+
+        Examples
+        --------
+        >>> df = DataFrame({'A': [1, 2, 3], 'B': [1, 1, 1]})
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: DataFrame -> pd.DataFrame

File: pandas/core/groupby.py [hunk 6]
```diff
@@ -3899,6 +3899,54 @@ def count(self):
...
+        dropna : boolean, default True
+            Don't include NaN in the counts.
+
+        Returns
+        -------
+        nunique: DataFrame
+
+        Examples
+        --------
+        >>> df = DataFrame({'id': ['spam', 'egg', 'egg', 'spam', 'ham', 'ham'],
```
Comment (jorisvandenbossche) [inline pandas/core/groupby.py]: DataFrame -> pd.DataFrame

File: pandas/core/groupby.py [hunk 7]
```diff
@@ -3899,6 +3899,54 @@ def count(self):
...
+        5   ham       5      y
+
+        >>> df.groupby('id').nunique()
+            id  value1  value2
+        id
+        egg    1       1       1
+        ham    1       1       2
+        spam   1       2       1
+
+        >>> df.groupby('id').filter(lambda g: (g.nunique() > 1).any())
```
Comment (jorisvandenbossche) [inline pandas/core/groupby.py]: This is a nice example (although more of the filter method and dataframe.nunique), but can you add one sentence introducing it? (explaining what we are going to do in the next example)

File: pandas/core/groupby.py [hunk 8]
```diff
@@ -3899,6 +3899,54 @@ def count(self):
...
+             id  value1 value2
+        0  spam       1      a
+        3  spam       2      a
+        4   ham       5      x
+        5   ham       5      y
+        """
+        from functools import partial
+        func = partial(Series.nunique, dropna=dropna)
+        return self.apply(lambda g: g.apply(func))
```
Comment (jorisvandenbossche) [inline pandas/core/groupby.py]: Is the use of `functools.partial` actually needed here? As extra keyword like dropna passed to `apply` will normally be passed through to the `func`

File: pandas/tests/frame/test_analytics.py [hunk 9]
```diff
@@ -410,6 +411,11 @@ def test_count(self):
         expected = Series(0, index=[])
         tm.assert_series_equal(result, expected)
 
+    def test_nunique(self):
+        f = lambda s: len(algorithms.unique1d(s.dropna()))
+        self._check_stat_op('nunique', f, has_skipna=False,
+                            check_dtype=False, check_dates=True)
+
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_analytics.py]: Can you add tests for the dropna and axis keywords as well?

File: pandas/core/frame.py [hunk 10]
```diff
@@ -4969,6 +4969,39 @@ def f(x):
...
+        """
+        Return Series with number of distinct observations over requested
+        axis.
+
+        .. versionadded:: 0.20.0
+
+        Parameters
+        ----------
+        axis : {0 or 'index', 1 or 'columns'}, default 0
+            0 or 'index' for row-wise, 1 or 'columns' for column-wise
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: row-wise / column-wise is probably used a lot in other methods here as well (can you check?), but, I find it in this case a bit confusing. As I would interpret 'column-wise' as "distinct observations for each column". And this is not correct, as that is the default of axis=0/'index'. So the axis=1 is more 'over/along the columns' 
But English is not my mother tongue. What do you think?

**Review by jorisvandenbossche — COMMENTED** [review]
@xflr6 Nice enhancement! I added some more comments.

**Comment by jorisvandenbossche** [comment]
@xflr6 As @jreback said, it is no problem that you do not get the benchmarks working for this PR. But, just to try to solve it: there seems to go something wrong in building the environment. Can you yourself build pandas master in a conda env?

File: pandas/core/frame.py (hunk 10 above)
Comment (xflr6) [inline pandas/core/frame.py]: As is, the wording is consistent with all the other methods such as `count()`: Wouldn't it be better to have a dedicated PR for that, in case all the `axis` docstrings are to be improved?

File: pandas/core/frame.py (hunk 5 above)
Comment (xflr6) [inline pandas/core/frame.py]: Changed (note that there are still some other example sections with bare `DataFrame`).

File: pandas/core/groupby.py (hunk 8 above)
Comment (xflr6) [inline pandas/core/groupby.py]: TIL about kwargs-passing in `apply()`, simplified both `nunique`-methods with that.

File: pandas/core/frame.py (hunk 10 above)
Comment (jreback) [inline pandas/core/frame.py]: @xflr6 yes could have a PR for improving these kinds of things in general (we already use shared_docs for this type of thing anyhow), so these are pretty general. Here is not as ``.nunique`` has separate doc-strings for Series/DataFrame, which is why @jorisvandenbossche is asking.

ok with actually fixing that (so this would hook into our more general doc-strings system).

Comment (jreback) [inline pandas/core/frame.py]: actually ``Series.nunique`` is defined in ``pandas.core.base`` (so its the *same* for ``Index``). But these could easily hook into the same doc-string system as I said above.

Comment (jorisvandenbossche) [inline pandas/core/frame.py]: Note that it is not fully consistent throughout frame.py as well. There are also some methods that explain this differently (eg `apply`, `mode`. `corrwith` actually switches the row and column-wise ("0 or 'index' to compute column-wise, 1 or 'columns' for row-wise")). 

The thing is also that the explanation can be different depending on what the function does I think.

File: pandas/core/frame.py (hunk 5 above)
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: yep, but we try to have the standards for *new* code a bit higher :-)

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche if you have any final comments.

File: pandas/core/frame.py (hunk 10 above)
Comment (jreback) [inline pandas/core/frame.py]: @xflr6 can you make the doc-string references to the axis consistent with other methods

e.g. example from another method.
```
        axis : {0 or 'index', 1 or 'columns'}, default 0
            Sort index/rows versus columns
```

I think you can simply drop the 2nd line of the axis parm

**Comment by jreback** [comment]
thanks @xflr6

**Comment by jorisvandenbossche** [comment]
One remaining remark is that I think row-wise / column-wise is a wrong explanation in this case.

**Comment by jreback** [comment]
@jorisvandenbossche I took that out on the merge, FYI
